OutOfSchool/OutOfSchool.DataAccess/Repository/EntityRepositorySoftDeleted.cs

```diff
@@ -0,0 +1,134 @@
+        {
+            query = query.Include(includeProperty);
+        }
+
+        return await query.ToListAsync().ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public override async Task<IEnumerable<TEntity>> GetByFilter(
+        Expression<Func<TEntity, bool>> predicate,
+        string includeProperties = "")
+    {
+        predicate = GetWhereExpression(predicate);
+        return await base.GetByFilter(predicate, includeProperties).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public override IQueryable<TEntity> GetByFilterNoTracking(
+        Expression<Func<TEntity, bool>> predicate,
+        string includeProperties = "")
+    {
+        predicate = GetWhereExpression(predicate);
+        return base.GetByFilterNoTracking(predicate, includeProperties);
+    }
+
+    /// <inheritdoc/>
+    public override Task<TEntity> GetById(TKey id) => dbSet.FirstOrDefaultAsync(x => !x.IsDeleted && x.Id.Equals(id));
+
+    /// <inheritdoc/>
+    public override Task<int> Count(Expression<Func<TEntity, bool>> where = null)
+    {
+        return base.Count(GetWhereExpression(where));
+    }
+
+    public override Task<bool> Any(Expression<Func<TEntity, bool>> where = null)
+    {
+        return base.Any(GetWhereExpression(where));
+    }
+
+    public override IQueryable<TEntity> Get(
+        int skip = 0,
+        int take = 0,
+        string includeProperties = "",
+        Expression<Func<TEntity, bool>> where = null,
+        Dictionary<Expression<Func<TEntity, object>>, SortDirection> orderBy = null,
+        bool asNoTracking = false)
+    {
+        return base.Get(skip, take, includeProperties, GetWhereExpression(where), orderBy, asNoTracking);
+    }
+
+    private Expression<Func<TEntity, bool>> GetWhereExpression(Expression<Func<TEntity, bool>> where)
+    {
+        if (where != null)
+        {
+            Expression<Func<TEntity, bool>> right = x => !x.IsDeleted;
+            where = Expression.Lambda<Func<TEntity, bool>>(
+                Expression.AndAlso(
+                    where.Body,
+                    new ExpressionParameterReplacer(right.Parameters, where.Parameters).Visit(right.Body)
+                    ),
+                where.Parameters);
+        }
+        else
+        {
+            where = x => !x.IsDeleted;
+        }
+
+        return where;
+    }
+
+    private sealed class ExpressionParameterReplacer : ExpressionVisitor
+    {
+        private IDictionary<ParameterExpression, ParameterExpression> ParameterReplacements { get; set; }
+
+        public ExpressionParameterReplacer(
+            IList<ParameterExpression> fromParameters,
+            IList<ParameterExpression> toParameters)
+        {
+            ParameterReplacements = new Dictionary<ParameterExpression, ParameterExpression>();
+
+            for (int i = 0; i != fromParameters.Count && i != toParameters.Count; i++)
+            { ParameterReplacements.Add(fromParameters[i], toParameters[i]); }
```

Comment:
(Fundamental code design) Please check suggested code. Maybe it's less verbose way to fill that dictionary

```suggestion
            ParameterReplacements = fromParameters.Zip(toParameters, (k, v) => new { k, v })
                .ToDictionary(x => x.k, x => x.v);
```

Reply:
Perfect. I checked it. I left code from the internet sample without refactoring. I forgot about zip. Thank you.

Reply:
Thanks for taking this variant, and NP, my code is not ideal. 
It is worse from performance perspective, but better (I hope) from cognitive perspective. 
I usually choose second option, and only in rare cases (when it is proven that we have a real bottleneck) write more performant code. Your mileage might vary, that's why I wasn't insisting, just asking for your opinion.